bionic/datatypes.py

```diff
@@ -194,62 +194,52 @@ class CaseKey(ImmutableMapping):
+    MISSING = MissingCaseKeyToken()
 
-    def __init__(self, name_value_token_triples):
-        values_by_name = {
-            name: value for name, value, token in name_value_token_triples
-        }
-        tokens_by_name = {
-            name: token for name, value, token in name_value_token_triples
-        }
+    def __init__(self, name_token_tuples):
+        tokens_by_name = {name: token for name, token in name_token_tuples}
 
         super(CaseKey, self).__init__(tokens_by_name)
-        self._nvt_triples = name_value_token_triples
-        self.values = values_by_name
+        self._name_token_tuples = name_token_tuples
         self.tokens = tokens_by_name
-        self.space = CaseKeySpace(list(values_by_name.keys()))
+        self.space = CaseKeySpace(list(tokens_by_name.keys()))
         self.missing_names = [
             name
-            for name, value, token in name_value_token_triples
+            for name, token in name_token_tuples
             # We don't want to say `value == MISSING` because `value` might override the
             # `==` operator to mean something different (like a Pandas DataFrame does).
-            if type(value) is MissingCaseKeyValue
+            if type(token) is MissingCaseKeyToken
         ]
         self.has_missing_values = len(self.missing_names) > 0
 
     def project(self, key_space):
         return CaseKey(
             [
-                (name, value, token)
-                for name, value, token in self._nvt_triples
+                (name, token)
+                for name, token in self._name_token_tuples
                 if name in key_space
             ]
         )
 
     def drop(self, key_space):
         return CaseKey(
             [
-                (name, value, token)
-                for name, value, token in self._nvt_triples
+                (name, token)
+                for name, token in self._name_token_tuples
                 if name not in key_space
             ]
         )
 
     def merge(self, other):
-        vt_pairs_by_name = {
-            name: (value, token) for name, value, token in self._nvt_triples
-        }
+        tokens_by_name = {name: token for name, token in self._name_token_tuples}
 
-        for name, value, token in other._nvt_triples:
-            if name in vt_pairs_by_name:
-                assert token == vt_pairs_by_name[name][1]
+        for name, token in other._name_token_tuples:
+            if name in tokens_by_name:
+                assert token == tokens_by_name[name]
             else:
-                vt_pairs_by_name[name] = value, token
+                tokens_by_name[name] = token
 
-        return CaseKey(
-            [(name, value, token) for name, (value, token) in vt_pairs_by_name.items()]
-        )
+        return CaseKey([(name, token) for name, token in tokens_by_name.items()])
 
     def __repr__(self):
         args_str = ", ".join(f"{name}={token}" for name, token in self.items())
```

Comment:
We use tokens for returning a token in `multi_index_from_case_keys` and for finding out `MISSING` values.

Reply:
As you noted, we will now be printing `None` instead of `<MISSING>` when the token is missing. Since this is the same thing as what we print when the actual value is `None`, maybe we should continue to print `<MISSING>` here?